flink-table/flink-table-planner-blink/src/main/scala/org/apache/flink/table/planner/delegation/PlannerBase.scala

```diff
@@ -254,28 +252,30 @@ abstract class PlannerBase(
     */
   protected def translateToPlan(execNodes: util.List[ExecNode[_, _]]): util.List[Transformation[_]]
 
-  private def getTableSink(objectIdentifier: ObjectIdentifier): Option[TableSink[_]] = {
-    JavaScalaConversionUtil.toScala(catalogManager.getTable(objectIdentifier)) match {
+  private def getTableSink(identifier: ObjectIdentifier): Option[(CatalogTable, TableSink[_])] = {
+    JavaScalaConversionUtil.toScala(catalogManager.getTable(identifier)) match {
       case Some(s) if s.isInstanceOf[ConnectorCatalogTable[_, _]] =>
-        JavaScalaConversionUtil
-          .toScala(s.asInstanceOf[ConnectorCatalogTable[_, _]].getTableSink)
+        val table = s.asInstanceOf[ConnectorCatalogTable[_, _]]
+        JavaScalaConversionUtil.toScala(table.getTableSink) match {
+          case Some(sink) => Some(table, sink)
+          case None => None
+        }
 
       case Some(s) if s.isInstanceOf[CatalogTable] =>
-
-        val catalog = catalogManager.getCatalog(objectIdentifier.getCatalogName)
-        val catalogTable = s.asInstanceOf[CatalogTable]
+        val catalog = catalogManager.getCatalog(identifier.getCatalogName)
+        val table = s.asInstanceOf[CatalogTable]
         if (catalog.isPresent && catalog.get().getTableFactory.isPresent) {
-          val objectPath = objectIdentifier.toObjectPath
+          val objectPath = identifier.toObjectPath
           val sink = TableFactoryUtil.createTableSinkForCatalogTable(
             catalog.get(),
-            catalogTable,
+            table,
             objectPath)
           if (sink.isPresent) {
-            return Option(sink.get())
+            return Option(table, sink.get())
           }
         }
-        val sinkProperties = catalogTable.toProperties
-        Option(TableFactoryService.find(classOf[TableSinkFactory[_]], sinkProperties)
+        val sinkProperties = table.toProperties
+        Option(table, TableFactoryService.find(classOf[TableSinkFactory[_]], sinkProperties)
```

Comment:
Could you explain what's the difference between `TableFactoryUtil.createTableSinkForCatalogTable` and `TableFactoryService.find(classOf[TableSinkFactory[_]], sinkProperties)`?

Reply:
See `Catalog.getTableFactory`.
Option 1 use `Catalog.getTableFactory` to get `TableFactory` and create sink.
Option 2 use `TableFactoryService` to create `TableFactory` and create sink.
The reason why we need option 1 is that hive table factory can not find by  `TableFactoryService`, but I think this can be improved in future.